### DISCUSSION TITLE
What does MagickRelinquishMemory do exactly?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5718] [Development] [answered]
I am writing bindings for imagemagick in dart, so bear with me a little, I'll have some questions about the non-obvious things in the api.

The docs for `MagickRelinquishMemory` says that it "relinquishes` memory used by methods like MagickGetException()". So I tried this:

 ```
	MagickWand* m = NewMagickWand();
	MagickReadImage(m, "some invalid path");
	ExceptionType* exceptionType = malloc(sizeof(ExceptionType));
	char* description = MagickGetException((const MagickWand*)m, exceptionType);
	printf("\nBefore relinquish: description = %s\n exceptionType = %d\n", description, *exceptionType);
	printf("\n Before relinquish: isMagickWand = %d\n", IsMagickWand(m)); //prints 1
        MagickRelinquishMemory(m);
	printf("\n After relinquish: isMagickWand = %d\n", IsMagickWand(m)); // prints 0
	description = MagickGetException((const MagickWand*)m, exceptionType); // expected description to be empty
	printf("\nAfter relinquish: description = %s\n exceptionType = %d\n", description, *exceptionType);
```

I expected `description` to be empty the second time and `exceptionType` to be `UndefinedException` (i.e. no exception), which means I expected the wand to be cleared from exceptions, but this was not the case, the wand still contains the same exceptions.

So I expected now that this method `free`s the wand, but then I used some methods on the wand (like MagickReadImage), and they still work normally (no crashes or memory violations...), so I don't understand what changed after calling this method, and what does it do excatly?

p.s.: I need to know because this method takes a pointer as  an arg, and in dart this will not be the case, so I want to decide how should I adapt it in dart.

## update

I found [this](https://legacy.imagemagick.org/discourse-server/viewtopic.php?f=6&t=7523) discussion on the legacy imagemagick website, which says that I should use it to free resources returned to me by methods like `MagickGetImageFilename()`, does this mean that I should use `MagickRelinquishMemory` instead of malloc's `free` method to release these kinds of resources?

**Comment by snibgo** [comment]
The source code magick-wand.c has a comment:
> NewMagickWand() returns a wand required for all other methods in the API. A fatal exception is thrown if there is not enough memory to allocate the wand. Use DestroyMagickWand() to dispose of the wand when it is no longer needed.

And DestroyMagickWand() calls RelinquishMagickMemory() among other operations.

But you call `MagickRelinquishMemory(m);` directly. This seems unwise.

**Reply by Haidar0096** [reply]
@snibgo this is not production code,this code was just to test what does Relinquish function do.

Do you know what does it do?
why should I use it, say for example, to free a char*, as recommended in the link mentioned above, instead of using free() function ?

**Comment by snibgo** [comment, marked as answer]
IM is open source, so you can read the code to see what it does.

In general, when IM has a function that creates an object (eg `NewMagickWand()`) it also provides a function to destroy that object (eg `DestroyMagickWand()`) that also frees up the memory. I suggest always using the appropriate destroy function.

If you have called `malloc()`, then using `free()` is reasonable. But don't use `free()` on IM structures that have been created by IM functions.

**Reply by Haidar0096** [reply]
@snibgo I understand really, but the provided link references an answer from an imagemagick site admin who recommends freeing a char* using Relinquish function, that is what I am asking about. I understand it is better yo use destroy over a wand, but destroy on char*? why?

**Comment by snibgo** [comment]
See the source code for `MagickGetException()`. This calls `AcquireQuantumMemory()`, so needs an IM function to free the memory. That function is `MagickRelinquishMemory()`. See `memory.c` for more details.

**Reply by Haidar0096** [reply]
@snibgo  I already saw the code and I know about AquireQuantumMemory. I think I am not making myself clear, this is my question:

Calling `char* e = MagickGetException(...)` returns a char*, are you telling me that `free(e)` is not equivalent to `MagickRelinquishMemory(e)`  in this case and will cause memory leaks?

**Comment by snibgo** [comment]
`free(e)` is not equivalent to `MagickRelinquishMemory(e)`.

**Reply by Haidar0096** [reply]
@snibgo thanks for clarification, so to put it all together, in an example:

```
    MagickWand *wand = NewMagickWand();
    MagickReadImage(wand, "invalidPath.png");
    ExceptionType *exceptionType = malloc(sizeof *exceptionType);
    char* description = MagickGetException(wand, exceptionType);
    RelinquishMagickMemory(description); // use Relinquish on pointers created by IM
    free(exceptionType); // use free on my defined pointers
```

**Comment by snibgo** [comment]
I suggest adding this at the end:
```
wand = DestroyMagickWand (wand);
```

**Reply by Haidar0096** [reply]
@snibgo of course, my snippet was just so you approve that freeing the variables (without destroying the wand) is done correctly.

I will mark your answer as accepted. Thanks again :)
